File: runtime/host/workbook.ex
defmodule Workbooks.Workbook do
  @moduledoc """
  HTML-native reader for workbooks. A workbook IS an HTML file built from the
  `work-*` Lit web components (the `workponents/` library). The backend never
  needs a bespoke format or kernel to understand it — it parses the HTML with a
  standard parser (Floki) and reads the `work-*` element tree.

  This replaces the old wasm kernel. The jobs the kernel did map onto plain HTML
  parsing:

    * `parse_headlines/1` — the structure/outline: every `work-*` element flattened
      to `{level, title, status, tags, props, lang, ...}` rows.
    * `tangle_plan/1` — the build plan: find `<work-component>` elements (their
      `lang`/`in`/`out`/`deps`/`uses`/`dir` attributes + text body = source) grouped
      under their `<work-flow>` worlds, so the compiler lane can build them.
    * `validate/1` — basic checks over the parsed component graph.
    * `render/1` — a workbook IS HTML; render is the HTML itself (passthrough). The
      browser + the `work-*` Lit components do the visual rendering.

  No GenServer, no wasm, no host calls — Floki over a binary, pure compute.
  """

  # Reserved attributes carry their own meaning (title/status/tags/build wiring)
  # and so are NOT surfaced as generic "props".
  @reserved ~w(id title status tag assignee lang in out deps uses persist dir
               due scheduled at start end cron repeat)

  @lang_ext %{
    "rust" => ".rs",
    "rs" => ".rs",
    "c" => ".c",
    "zig" => ".zig",
    "javascript" => ".js",
    "js" => ".js",
    "typescript" => ".ts",
    "ts" => ".ts"
  }

  @doc "Extract the Workbook Spec (JSON) from artifact HTML."
  def spec(html) when is_binary(html) do
    case Regex.run(~r/<script id="workbook-spec"[^>]*>(.*?)<\/script>/s, html) do
      [_, json] -> Jason.decode(json)
      _ -> {:error, :no_spec}
    end
  end

  @doc """
  Parse a workbook's HTML → a flat list of `work-*` node rows in document order.
  Each row: `%{"level", "title", "status", "tags", "props", "id", "lang", "in",
  "out", "deps", "uses", "dir"}`. `level` is 1-based depth in the `work-*`
  element tree.
  """
  def parse_headlines(src) when is_binary(src) do
    src
    |> parse_work_nodes()
    |> flatten(1)
    |> Enum.map(&node_row/1)
  end

  @doc "Lint a workbook → list of diagnostics (currently none; kept for parity)."
  def lint(_src), do: []

  @doc """
  Emit the build plan from a workbook: `%{"worlds" => [...]}`. Each top-level
  `<work-flow>` is a world; its descendant `<work-component>` elements are the
  buildable components (source = the element's text body, wiring from its
  attributes). A document with no `<work-flow>` but with bare `<work-component>`
  elements yields a single implicit world so loose components still tangle/build.
  """
  def tangle_plan(src) when is_binary(src) do
    roots = parse_work_nodes(src)

    worlds =
      case top_flows(roots) do
        [] ->
          comps = all_components(roots)
          if comps == [], do: [], else: [build_world(implicit_flow(comps))]

        flows ->
          Enum.map(flows, &build_world/1)
      end

    %{"worlds" => worlds}
  end

  @doc """
  Validate a workbook → list of diagnostics. Checks every `work-component` in a
  `work-flow`: a component must declare a language, and any `in` it consumes must
  have an upstream producer (`out`) in the same flow.
  """
  def validate(src) when is_binary(src) do
    roots = parse_work_nodes(src)

    flows =
      case top_flows(roots) do
        [] ->
          comps = all_components(roots)
          if comps == [], do: [], else: [implicit_flow(comps)]

        fs ->
          fs
      end

    Enum.flat_map(flows, fn flow ->
      comps = comps_of(flow)
      produced = comps |> Enum.map(& &1.out) |> Enum.reject(&is_nil/1) |> MapSet.new()

      Enum.flat_map(comps, fn c ->
        []
        |> then(fn d ->
          if c.lang == nil,
            do: [diag("error", c.name, "component has no source block / language") | d],
            else: d
        end)
        |> then(fn d ->
          if c.inp && not MapSet.member?(produced, c.inp),
            do: [diag("error", c.name, "input `#{c.inp}` has no upstream producer") | d],
            else: d
        end)
      end)
    end)
  end

  @doc "The native file extension for a component language, or nil if unbuildable."
  def lang_ext(lang) when is_binary(lang), do: @lang_ext[String.downcase(lang)]
  def lang_ext(_), do: nil

  # ── HTML → work-* node tree (Floki) ──────────────────────────────────────

  # A node mirrors the old kernel's Node: tag + attrs + body text + children.
  defp parse_work_nodes(src) do
    case Floki.parse_fragment(src) do
      {:ok, tree} -> Enum.flat_map(tree, &to_nodes/1)
      {:error, _} -> []
    end
  end

  # Keep only `work-*` elements as nodes; descend through non-work wrappers so a
  # `work-*` element nested inside plain HTML is still found.
  defp to_nodes({tag, attrs, children}) when is_binary(tag) do
    if String.starts_with?(tag, "work-") do
      [
        %{
          tag: tag,
          attrs: Map.new(attrs),
          body: own_text(children),
          children: Enum.flat_map(children, &to_nodes/1)
        }
      ]
    else
      Enum.flat_map(children, &to_nodes/1)
    end
  end

  defp to_nodes(_), do: []

  # Direct text of an element (its source body), trimmed, excluding child elements.
  defp own_text(children) do
    children
    |> Enum.filter(&is_binary/1)
    |> Enum.join()
    |> String.trim()
  end

  defp flatten(nodes, level) do
    Enum.flat_map(nodes, fn n ->
      [{level, n} | flatten(n.children, level + 1)]
    end)
  end

  # ── Row shape for parse_headlines ────────────────────────────────────────

  defp node_row({level, n}) do
    attrs = n.attrs
    status = attrs["status"] && String.upcase(attrs["status"])

    %{
      "level" => level,
      "title" => attrs["title"] || "",
      "status" => status,
      "state" => status,
      "id" => attrs["id"],
      "tag" => attrs["tag"],
      "tags" => tags_of(attrs),
      "props" => props_of(attrs),
      "lang" => attrs["lang"],
      "in" => attrs["in"],
      "out" => attrs["out"],
      "dir" => attrs["dir"],
      "tagname" => n.tag
    }
  end

  defp tags_of(attrs) do
    case attrs["tag"] do
      t when is_binary(t) -> String.split(t)
      _ -> []
    end
  end

  defp props_of(attrs) do
    attrs
    |> Enum.reject(fn {k, _} -> k in @reserved end)
    |> Map.new(fn {k, v} -> {String.upcase(k), v} end)
  end

  # ── Components + build plan ───────────────────────────────────────────────

  defp component?(n), do: n.tag == "work-component"
  defp flow?(n), do: n.tag == "work-flow"

  # The top-most work-flow in each branch; recursion stops at the first flow.
  defp top_flows(nodes) do
    Enum.flat_map(nodes, fn n ->
      if flow?(n), do: [n], else: top_flows(n.children)
    end)
  end

  # Every work-component anywhere in the tree (for the implicit-world fallback).
  defp all_components(nodes) do
    Enum.flat_map(nodes, fn n ->
      child = all_components(n.children)
      if component?(n), do: [n | child], else: child
    end)
  end

  defp implicit_flow(component_nodes) do
    %{tag: "work-flow", attrs: %{}, body: "", children: component_nodes}
  end

  # A component descriptor read off a work-component element.
  defp comp_of(n) do
    a = n.attrs

    %{
      name: a["title"] || "",
      # `lang` present (even empty) = a declared source block.
      lang: a["lang"],
      deps: list_attr(a, "deps"),
      uses: list_attr(a, "uses"),
      inp: nonempty(a["in"]),
      out: nonempty(a["out"]),
      persist: Map.has_key?(a, "persist"),
      src: n.body,
      dir: nonempty(a["dir"])
    }
  end

  defp comps_of(flow) do
    Enum.flat_map(flow.children, fn n ->
      child = comps_of(n)
      if component?(n), do: [comp_of(n) | child], else: child
    end)
  end

  defp build_world(flow) do
    {comps, subworlds} =
      Enum.reduce(flow.children, {[], []}, fn child, {cs, ws} ->
        cond do
          flow?(child) -> {cs, [build_world(child) | ws]}
          component?(child) -> {[comp_of(child) | cs], ws}
          true -> {cs, ws}
        end
      end)

    comps = Enum.reverse(comps)
    subworlds = Enum.reverse(subworlds)
    {imports, exports, edges} = sig_of(comps)

    %{
      "name" => flow.attrs["title"] || "",
      "schedule" => flow.attrs["cron"] || flow.attrs["scheduled"] || flow.attrs["at"],
      "imports" => imports,
      "exports" => exports,
      "components" => Enum.map(comps, &comp_json/1),
      "edges" => Enum.map(edges, fn {f, t} -> %{"from" => f, "to" => t} end),
      "workflows" => subworlds
    }
  end

  defp sig_of(comps) do
    imports = comps |> Enum.flat_map(& &1.uses) |> Enum.sort() |> Enum.dedup()

    producer =
      Enum.reduce(comps, %{}, fn c, acc ->
        if c.out, do: Map.put(acc, c.out, c.name), else: acc
      end)

    {edges, consumed} =
      Enum.reduce(comps, {[], MapSet.new()}, fn c, {es, cons} ->
        case c.inp && producer[c.inp] do
          nil -> {es, cons}
          from -> {[{from, c.name} | es], MapSet.put(cons, c.inp)}
        end
      end)

    exports =
      comps
      |> Enum.map(& &1.out)
      |> Enum.reject(&is_nil/1)
      |> Enum.reject(&MapSet.member?(consumed, &1))
      |> Enum.sort()
      |> Enum.dedup()

    {imports, exports, Enum.reverse(edges)}
  end

  defp comp_json(c) do
    %{
      "name" => c.name,
      "lang" => c.lang,
      "deps" => c.deps,
      "uses" => c.uses,
      "in" => c.inp,
      "out" => c.out,
      "persist" => c.persist,
      "src" => c.src,
      "dir" => c.dir
    }
  end

  # ── helpers ───────────────────────────────────────────────────────────────

  defp list_attr(attrs, key) do
    case attrs[key] do
      v when is_binary(v) and v != "" ->
        v |> String.split(",") |> Enum.map(&String.trim/1) |> Enum.reject(&(&1 == ""))

      _ ->
        []
    end
  end

  defp nonempty(v) when is_binary(v) and v != "", do: v
  defp nonempty(_), do: nil

  defp diag(level, scope, msg), do: %{"level" => level, "scope" => scope, "message" => msg}
end

File: cli/src/workbook.rs
//! HTML-native workbook reader (CLI side, dep-free).
//!
//! A workbook IS an HTML file built from `work-*` web components. The CLI reads
//! its STRUCTURE — the `work-*` element tree — with a small standard-shaped HTML
//! scanner (no kernel, no wasm). It mirrors `Workbooks.Workbook` (Elixir/Floki):
//!
//!   * `parse_headlines` : the outline — every `work-*` element as a row.
//!   * `tangle_plan`     : the build plan — `<work-flow>` worlds + `<work-component>`
//!                         leaves (source = element body, wiring = attributes).
//!   * `validate`        : diagnostics over the parsed component graph.
//!   * `render`          : a workbook IS HTML; render is the HTML itself (passthrough).
//!
//! The scanner only needs to recognise `work-*` tags; everything else is opaque
//! markup it descends through. Pure compute, identical on native + wasm targets.

use serde_json::{json, Value};
use std::collections::BTreeMap;

const RESERVED: &[&str] = &[
    "id", "title", "status", "tag", "assignee", "lang", "in", "out", "deps", "uses", "persist",
    "dir", "due", "scheduled", "at", "start", "end", "cron", "repeat",
];

#[derive(Debug, Clone)]
struct Node {
    tag: String,
    attrs: BTreeMap<String, String>,
    body: String,
    children: Vec<Node>,
}

// ── Public API (JSON strings out, mirroring the Elixir surface) ────────────

pub fn parse_headlines(src: &str) -> String {
    let roots = parse_work_nodes(src);
    let mut rows = vec![];
    for n in &roots {
        flatten(n, 1, &mut rows);
    }
    let arr: Vec<Value> = rows.iter().map(|(level, n)| node_row(*level, n)).collect();
    serde_json::to_string(&arr).unwrap_or_else(|_| "[]".into())
}

// Diagnostics that aren't structural validation (currently none). The `lint`
// verb uses `validate`; kept for API parity with the Elixir surface.
#[allow(dead_code)]
pub fn lint(_src: &str) -> String {
    "[]".into()
}

pub fn tangle_plan(src: &str) -> String {
    let roots = parse_work_nodes(src);
    let worlds: Vec<Value> = match top_flows(&roots) {
        flows if !flows.is_empty() => flows.iter().map(|f| build_world(f)).collect(),
        _ => {
            let comps = all_components(&roots);
            if comps.is_empty() {
                vec![]
            } else {
                vec![build_world(&implicit_flow(comps))]
            }
        }
    };
    json!({ "worlds": worlds }).to_string()
}

pub fn validate(src: &str) -> String {
    let roots = parse_work_nodes(src);
    let flows: Vec<Node> = match top_flows(&roots) {
        flows if !flows.is_empty() => flows.into_iter().cloned().collect(),
        _ => {
            let comps = all_components(&roots);
            if comps.is_empty() {
                vec![]
            } else {
                vec![implicit_flow(comps)]
            }
        }
    };

    let mut diags = vec![];
    for flow in &flows {
        let comps = comps_of(flow);
        let produced: std::collections::BTreeSet<&str> =
            comps.iter().filter_map(|c| c.out.as_deref()).collect();
        for c in &comps {
            if c.lang.is_none() {
                diags.push(diag("error", &c.name, "component has no source block / language"));
            }
            if let Some(inp) = &c.inp {
                if !produced.contains(inp.as_str()) {
                    diags.push(diag(
                        "error",
                        &c.name,
                        &format!("input `{}` has no upstream producer", inp),
                    ));
                }
            }
        }
    }
    serde_json::to_string(&diags).unwrap_or_else(|_| "[]".into())
}

/// A workbook IS HTML — render returns the HTML itself (passthrough). The browser
/// + the `work-*` Lit components do the visual render.
pub fn render(src: &str) -> String {
    src.to_string()
}

// ── HTML → work-* node tree (minimal scanner) ──────────────────────────────

fn parse_work_nodes(src: &str) -> Vec<Node> {
    let toks = tokenize(src);
    let mut pos = 0;
    let mut roots = vec![];
    while pos < toks.len() {
        if let Some((node, next)) = parse_node(&toks, pos) {
            if node.tag.starts_with("work-") {
                roots.push(node);
            } else {
                // descend: a non-work element may wrap work-* children.
                roots.extend(node.children);
            }
            pos = next;
        } else {
            pos += 1;
        }
    }
    roots
}

#[derive(Debug, Clone)]
enum Tok {
    Open(String, BTreeMap<String, String>, bool), // tag, attrs, self_closing
    Close(String),
    Text(String),
}

fn tokenize(src: &str) -> Vec<Tok> {
    let bytes: Vec<char> = src.chars().collect();
    let mut i = 0;
    let mut out = vec![];
    let mut text = String::new();

    while i < bytes.len() {
        if bytes[i] == '<' {
            // flush pending text
            if !text.is_empty() {
                out.push(Tok::Text(std::mem::take(&mut text)));
            }
            // find the matching '>'
            let start = i;
            let mut j = i + 1;
            let mut in_q: Option<char> = None;
            while j < bytes.len() {
                let c = bytes[j];
                match in_q {
                    Some(q) if c == q => in_q = None,
                    Some(_) => {}
                    None if c == '"' || c == '\'' => in_q = Some(c),
                    None if c == '>' => break,
                    None => {}
                }
                j += 1;
            }
            if j >= bytes.len() {
                // unterminated tag — treat the rest as text and stop.
                text.extend(&bytes[start..]);
                break;
            }
            let raw: String = bytes[start + 1..j].iter().collect();
            i = j + 1;
            if let Some(tok) = parse_tag(&raw) {
                out.push(tok);
            }
            // comments / doctype / processing instrs → dropped (parse_tag → None)
        } else {
            text.push(bytes[i]);
            i += 1;
        }
    }
    if !text.is_empty() {
        out.push(Tok::Text(text));
    }
    out
}

fn parse_tag(raw: &str) -> Option<Tok> {
    let raw = raw.trim();
    if raw.is_empty() || raw.starts_with('!') || raw.starts_with('?') {
        return None;
    }
    if let Some(rest) = raw.strip_prefix('/') {
        return Some(Tok::Close(rest.trim().to_lowercase()));
    }
    let self_closing = raw.ends_with('/');
    let body = raw.trim_end_matches('/').trim();

    // tag name = up to first whitespace
    let (name, attr_str) = match body.find(|c: char| c.is_whitespace()) {
        Some(idx) => (&body[..idx], &body[idx..]),
        None => (body, ""),
    };
    let attrs = parse_attrs(attr_str);
    Some(Tok::Open(name.to_lowercase(), attrs, self_closing))
}

fn parse_attrs(s: &str) -> BTreeMap<String, String> {
    let chars: Vec<char> = s.chars().collect();
    let mut i = 0;
    let mut attrs = BTreeMap::new();
    while i < chars.len() {
        while i < chars.len() && chars[i].is_whitespace() {
            i += 1;
        }
        if i >= chars.len() {
            break;
        }
        // read name
        let nstart = i;
        while i < chars.len() && chars[i] != '=' && !chars[i].is_whitespace() {
            i += 1;
        }
        let name: String = chars[nstart..i].iter().collect::<String>().to_lowercase();
        if name.is_empty() {
            i += 1;
            continue;
        }
        // optional = value
        while i < chars.len() && chars[i].is_whitespace() {
            i += 1;
        }
        if i < chars.len() && chars[i] == '=' {
            i += 1;
            while i < chars.len() && chars[i].is_whitespace() {
                i += 1;
            }
            let val = if i < chars.len() && (chars[i] == '"' || chars[i] == '\'') {
                let q = chars[i];
                i += 1;
                let vstart = i;
                while i < chars.len() && chars[i] != q {
                    i += 1;
                }
                let v: String = chars[vstart..i].iter().collect();
                if i < chars.len() {
                    i += 1; // closing quote
                }
                v
            } else {
                let vstart = i;
                while i < chars.len() && !chars[i].is_whitespace() {
                    i += 1;
                }
                chars[vstart..i].iter().collect()
            };
            attrs.insert(name, html_unescape(&val));
        } else {
            // boolean attribute
            attrs.insert(name, String::new());
        }
    }
    attrs
}

fn html_unescape(s: &str) -> String {
    s.replace("&lt;", "<")
        .replace("&gt;", ">")
        .replace("&quot;", "\"")
        .replace("&#39;", "'")
        .replace("&amp;", "&")
}

// Recursive-descent over the flat token stream → a Node (and the next index).
fn parse_node(toks: &[Tok], pos: usize) -> Option<(Node, usize)> {
    match toks.get(pos)? {
        Tok::Open(tag, attrs, self_closing) => {
            let mut node = Node {
                tag: tag.clone(),
                attrs: attrs.clone(),
                body: String::new(),
                children: vec![],
            };
            if *self_closing || is_void(tag) {
                return Some((node, pos + 1));
            }
            let mut i = pos + 1;
            while i < toks.len() {
                match &toks[i] {
                    Tok::Close(close) if close == tag => {
                        return Some((node, i + 1));
                    }
                    Tok::Close(_) => {
                        // stray/mismatched close — skip it.
                        i += 1;
                    }
                    Tok::Text(t) => {
                        node.body.push_str(t);
                        i += 1;
                    }
                    Tok::Open(..) => {
                        if let Some((child, next)) = parse_node(toks, i) {
                            if child.tag.starts_with("work-") {
                                node.children.push(child);
                            } else {
                                // flatten non-work descendants up so nested work-*
                                // elements are still discovered.
                                node.children.extend(child.children);
                            }
                            i = next;
                        } else {
                            i += 1;
                        }
                    }
                }
            }
            // unterminated element — accept what we have.
            Some((node, i))
        }
        _ => None,
    }
}

fn is_void(tag: &str) -> bool {
    matches!(
        tag,
        "area" | "base" | "br" | "col" | "embed" | "hr" | "img" | "input" | "link" | "meta"
            | "param" | "source" | "track" | "wbr"
    )
}

// ── Row shape for parse_headlines ──────────────────────────────────────────

fn flatten<'a>(n: &'a Node, level: usize, out: &mut Vec<(usize, &'a Node)>) {
    out.push((level, n));
    for c in &n.children {
        flatten(c, level + 1, out);
    }
}

fn node_row(level: usize, n: &Node) -> Value {
    let status = n.attrs.get("status").map(|s| s.to_uppercase());
    let tags: Vec<&str> = n
        .attrs
        .get("tag")
        .map(|t| t.split_whitespace().collect())
        .unwrap_or_default();
    let props: BTreeMap<String, &String> = n
        .attrs
        .iter()
        .filter(|(k, _)| !RESERVED.contains(&k.as_str()))
        .map(|(k, v)| (k.to_uppercase(), v))
        .collect();

    json!({
        "level": level,
        "title": n.attrs.get("title").cloned().unwrap_or_default(),
        "status": status,
        "state": status,
        "id": n.attrs.get("id"),
        "tag": n.attrs.get("tag"),
        "tags": tags,
        "props": props,
        "lang": n.attrs.get("lang"),
        "in": n.attrs.get("in"),
        "out": n.attrs.get("out"),
        "dir": n.attrs.get("dir"),
        "tagname": n.tag,
    })
}

// ── Components + build plan ─────────────────────────────────────────────────

struct Comp {
    name: String,
    lang: Option<String>,
    deps: Vec<String>,
    uses: Vec<String>,
    inp: Option<String>,
    out: Option<String>,
    persist: bool,
    src: String,
    dir: Option<String>,
}

fn is_flow(n: &Node) -> bool {
    n.tag == "work-flow"
}
fn is_component(n: &Node) -> bool {
    n.tag == "work-component"
}

fn top_flows(nodes: &[Node]) -> Vec<&Node> {
    let mut out = vec![];
    for n in nodes {
        if is_flow(n) {
            out.push(n);
        } else {
            out.extend(top_flows(&n.children));
        }
    }
    out
}

fn all_components(nodes: &[Node]) -> Vec<Node> {
    let mut out = vec![];
    for n in nodes {
        if is_component(n) {
            out.push(n.clone());
        }
        out.extend(all_components(&n.children));
    }
    out
}

fn implicit_flow(components: Vec<Node>) -> Node {
    Node {
        tag: "work-flow".into(),
        attrs: BTreeMap::new(),
        body: String::new(),
        children: components,
    }
}

fn list_attr(n: &Node, key: &str) -> Vec<String> {
    match n.attrs.get(key) {
        Some(v) if !v.is_empty() => v.split(',').map(|s| s.trim().to_string()).collect(),
        _ => vec![],
    }
}

fn opt_attr(n: &Node, key: &str) -> Option<String> {
    n.attrs.get(key).filter(|v| !v.is_empty()).cloned()
}

fn comp_of(n: &Node) -> Comp {
    Comp {
        name: n.attrs.get("title").cloned().unwrap_or_default(),
        lang: n.attrs.get("lang").cloned(),
        deps: list_attr(n, "deps"),
        uses: list_attr(n, "uses"),
        inp: opt_attr(n, "in"),
        out: opt_attr(n, "out"),
        persist: n.attrs.contains_key("persist"),
        src: n.body.trim().to_string(),
        dir: opt_attr(n, "dir"),
    }
}

fn comps_of(flow: &Node) -> Vec<Comp> {
    fn go(nodes: &[Node], out: &mut Vec<Comp>) {
        for n in nodes {
            if is_component(n) {
                out.push(comp_of(n));
            }
            go(&n.children, out);
        }
    }
    let mut out = vec![];
    go(&flow.children, &mut out);
    out
}

fn build_world(flow: &Node) -> Value {
    let mut comps = vec![];
    let mut subworlds = vec![];
    for child in &flow.children {
        if is_flow(child) {
            subworlds.push(build_world(child));
        } else if is_component(child) {
            comps.push(comp_of(child));
        }
    }
    let (imports, exports, edges) = sig_of(&comps);
    let components: Vec<Value> = comps.iter().map(comp_json).collect();
    let edges: Vec<Value> = edges
        .iter()
        .map(|(f, t)| json!({ "from": f, "to": t }))
        .collect();
    json!({
        "name": flow.attrs.get("title").cloned().unwrap_or_default(),
        "imports": imports,
        "exports": exports,
        "components": components,
        "edges": edges,
        "workflows": subworlds,
    })
}

fn sig_of(comps: &[Comp]) -> (Vec<String>, Vec<String>, Vec<(String, String)>) {
    let mut imports: Vec<String> = comps.iter().flat_map(|c| c.uses.clone()).collect();
    imports.sort();
    imports.dedup();

    let mut producer: BTreeMap<String, String> = BTreeMap::new();
    for c in comps {
        if let Some(o) = &c.out {
            producer.insert(o.clone(), c.name.clone());
        }
    }
    let mut edges = vec![];
    let mut consumed = std::collections::BTreeSet::new();
    for c in comps {
        if let Some(i) = &c.inp {
            if let Some(f) = producer.get(i) {
                edges.push((f.clone(), c.name.clone()));
                consumed.insert(i.clone());
            }
        }
    }
    let mut exports: Vec<String> = comps
        .iter()
        .filter_map(|c| c.out.clone())
        .filter(|o| !consumed.contains(o))
        .collect();
    exports.sort();
    exports.dedup();
    (imports, exports, edges)
}

fn comp_json(c: &Comp) -> Value {
    json!({
        "name": c.name, "lang": c.lang, "deps": c.deps, "uses": c.uses,
        "in": c.inp, "out": c.out, "persist": c.persist, "src": c.src, "dir": c.dir,
    })
}

fn diag(level: &str, scope: &str, msg: &str) -> Value {
    json!({ "level": level, "scope": scope, "message": msg })
}

#[cfg(test)]
mod tests {
    use super::*;

    const FLOW: &str = r#"
<work-flow title="etl">
  <work-component title="extract" lang="rust" out="raw"></work-component>
  <work-component title="transform" lang="rust" in="raw" out="clean" uses="fetch"></work-component>
  <work-component title="load" lang="rust" in="clean"></work-component>
</work-flow>
"#;

    #[test]
    fn parse_headlines_shape() {
        let json: Value = serde_json::from_str(&parse_headlines(FLOW)).unwrap();
        let rows = json.as_array().unwrap();
        assert_eq!(rows.len(), 4);
        assert_eq!(rows[0]["level"], 1);
        assert_eq!(rows[0]["title"], "etl");
        assert_eq!(rows[1]["level"], 2);
        assert_eq!(rows[1]["title"], "extract");
    }

    #[test]
    fn tangle_plan_world_and_edges() {
        let plan: Value = serde_json::from_str(&tangle_plan(FLOW)).unwrap();
        let worlds = plan["worlds"].as_array().unwrap();
        assert_eq!(worlds.len(), 1);
        let w = &worlds[0];
        assert_eq!(w["name"], "etl");
        assert_eq!(w["components"].as_array().unwrap().len(), 3);
        assert_eq!(w["imports"], json!(["fetch"]));
        let edges = w["edges"].as_array().unwrap();
        assert!(edges
            .iter()
            .any(|e| e["from"] == "extract" && e["to"] == "transform"));
    }

    #[test]
    fn validate_flags_dangling_and_langless() {
        let src = r#"<work-flow title="b">
          <work-component title="needs" lang="js" in="missing"></work-component>
          <work-component title="nolang"></work-component>
        </work-flow>"#;
        let diags: Value = serde_json::from_str(&validate(src)).unwrap();
        let msgs: Vec<String> = diags
            .as_array()
            .unwrap()
            .iter()
            .map(|d| d["message"].as_str().unwrap().to_string())
            .collect();
        assert!(msgs.iter().any(|m| m.contains("no upstream producer")));
        assert!(msgs.iter().any(|m| m.contains("no source block")));
    }

    #[test]
    fn validate_clean_is_empty() {
        assert_eq!(
            validate(r#"<work-doc title="t"><p>x</p></work-doc>"#),
            "[]"
        );
    }

    #[test]
    fn render_is_passthrough() {
        let html = r#"<work-doc title="Hi"><p>body</p></work-doc>"#;
        assert_eq!(render(html), html);
    }

    #[test]
    fn component_body_is_source() {
        let src = r#"<work-flow title="f"><work-component title="add" lang="rust" out="sum" dir="crates/add">pub fn add() {}</work-component></work-flow>"#;
        let plan: Value = serde_json::from_str(&tangle_plan(src)).unwrap();
        let comp = &plan["worlds"][0]["components"][0];
        assert_eq!(comp["name"], "add");
        assert_eq!(comp["dir"], "crates/add");
        assert!(comp["src"].as_str().unwrap().contains("pub fn add"));
    }

    #[test]
    fn descends_through_non_work_wrappers() {
        let json: Value =
            serde_json::from_str(&parse_headlines(r#"<div><work-note title="Inner"></work-note></div>"#))
                .unwrap();
        let rows = json.as_array().unwrap();
        assert_eq!(rows.len(), 1);
        assert_eq!(rows[0]["title"], "Inner");
        assert_eq!(rows[0]["level"], 1);
    }
}
